### PR TITLE
Embed only a subset of metav1.ObjectMeta fields

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
This PR removes all references to `metav1.ObjectMeta` to embedded resources. This patch came from a discussion in the [upstream kubebuilder slack](https://kubernetes.slack.com/archives/CAR30FCJZ/p1561092115400800). The `metav1.ObjectMeta` type, when embedded, contains a lot of fields that cannot be set from a client. This PR bring a copy of ObjectMeta into Cluster API including only fields that a user can set.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 1]
```diff
@@ -56,3 +57,69 @@ type MachineClassRef struct {
 	// +optional
 	Provider string `json:"provider,omitempty"`
 }
+
+// ObjectMeta is metadata that all persisted resources must have, which includes all objects
```
Comment (davidewatson) [inline pkg/apis/cluster/v1alpha1/common_types.go]: Can we leave more context? For example, why is this necessary for `Machine` and not other CRDs? And if it's only because `Machine` is embedded in other CRDs, then is there an upstream ticket to track this issue?

Comment (vincepri) [reply pkg/apis/cluster/v1alpha1/common_types.go]: Added more context

Comment (davidewatson) [reply pkg/apis/cluster/v1alpha1/common_types.go]: This is a fair explanation. I was mostly thinking about the problem this attempts to solve, and how it is likely to come up again. For example, suppose someone embeds `MachineDeployment` in another CRD. Won't they run into these same issues? And if so, why not remove `ObjectMeta` from `MachineDeployment` too?

I guess I am wondering what the long term guidance is from `apimachinery`...

Comment (ncdc) [reply pkg/apis/cluster/v1alpha1/common_types.go]: There's a couple of things going on here:

1. All objects have root-level `ObjectMeta`
1. Some objects, such as `Machine` and `MachineSet`, include non-root `ObjectMeta` somewhere in their nested structures

If anyone includes any Kubernetes object (built-in or CRD) in their own object, you'll end up with a similar situation: root-level `ObjectMetadata` along with non-root `ObjectMeta`. In the past, we've tried to design the API types such that the portions that someone might want to include are reusable (e.g. including `MachineSpec` in `MachineSet` (instead of a full `Machine`)). This could work, in theory, except some of these types have an `ObjectMeta` in them, even though they're not root-level types. It would be nice if these types used some other structure instead of `ObjectMeta`.

That's not necessarily a problem - we've been doing it for a while in CAPI, and core Kubernetes does it too. The problem @vincepri is running into is that the newer version of kubebuilder/controller-runtime/controller-tools generates openapi schema validation definitions that make it so you can't successfully create a Machine because of this nested `ObjectMeta` and a null `creationTimestamp` field.

Comment (ncdc) [reply pkg/apis/cluster/v1alpha1/common_types.go]: Also, when including a non-root `ObjectMeta` field, a lot of the time it's just to use it for the `annotations` and/or `labels`, but not the entire set of fields. It may be better to scope the non-root field to just the fields you need, instead of a full `ObjectMeta` struct.

Comment (neolit123) [reply pkg/apis/cluster/v1alpha1/common_types.go]: +1 on inspecting long term effects.

i think the PR should include more detail on the actual problem it's solving.
e.g. comments about:
```
validation failure list:
            spec.metadata.creationTimestamp in body must be of type string: "null"
```


something puzzling in the slack thread was that this started happening suddenly, did you figure out what caused that? was it https://github.com/kubernetes/kubernetes/pull/77653 ?

**Comment by davidewatson** [comment]
/hold cancel

Merging this now unblocks v1alpha2 work. We should discuss further on Wednesday.

**Comment by ncdc** [comment]
I like it! Merge, unblock, correct later if needed. 🎉

**Comment by vincepri** [comment]
@ncdc Adding more details to the commit, one second

**Comment by vincepri** [comment]
@ncdc @neolit123 @davidewatson Added more details to the comment above ObjectMeta.

@neolit123 The issue popped up for multiple reasons, I'm unaware of the implications for the PR you mentioned. The change that broke this behavior for us is in `controller-tools@v2`, which now doesn't add an exception for ObjectMeta and generates validation properties for all the sub-fields.

File: pkg/apis/cluster/v1alpha1/common_types.go [hunk 2]
```diff
@@ -56,3 +57,87 @@ type MachineClassRef struct {
...
+// specifically `spec.metadata.creationTimestamp in body must be of type string: "null"`.
+// The investigation showed that `controller-tools@v2` behaves differently than its previous version
+// when handling types from [metav1](k8s.io/apimachinery/pkg/apis/meta/v1) package.
+//
+// In more details, we found that embedded (non-top level) types that embedded `metav1.ObjectMeta`
+// had validation properties, including for `creationTimestamp` (metav1.Time).
+// The `metav1.Time` type specifies a custom json marshaller that, when IsZero() is true, returns `null`
+// which breaks validation because the field isn't marked as nullable.
+//
+// In future versions, controller-tools@v2 might allow ovveriding the type and validation for embedded
```
Comment (ncdc) [inline pkg/apis/cluster/v1alpha1/common_types.go]: typo: ovveriding
